focolab/minimo
# Issue 35

### Purpose ###
Disable deletion and overwrites of MinIO data. Unblock https://github.com/openjournals/joss-reviews/issues/2979.

### Approach ###
This is a bit ugly. MinIO offers [a simple solution](https://docs.min.io/docs/minio-bucket-object-lock-guide.html), but it only works for erasure coded deployments (i.e. not for deployments operating in Gateway mode). Moreover, it only works for buckets which have object locking enabled, and object locking can only be enabled at the time of initial bucket creation (i.e. not for buckets which already exist). MinIO's restrictions here follow Amazon's for S3, and accordingly, they are unlikely to change unless and until AWS enables object locking configuration for existing buckets.

With all of that in mind, I've implemented a WORM workaround according to the following approach:
1. Add a MinIO Client service definition to our Compose file. At time of application startup, this client creates a new user with separate credentials than those of the admin user. It also creates a new permissions policy which grants only the permissions for downloading existing files and viewing them in the data browser, and it assigns this policy to the new user.
2. Update the autologin script for the data browser to use the readonly user credentials. This should prevent overwrites and deletions from the data browser GUI.
3. Update our presigned URL endpoint to generate unique upload prefixes and prepend them in the URLs that it returns. This should guarantee that each data upload from the submit data page gets a unique prefix, which will prevent overwrites for successive uploads of identically named files.

### Testing ###
1. Build these changes locally with `docker-compose up -d --build --force-recreate`
2. Open web app in incognito tab, go to data browser, and verify that autologin works
3. Use the data browser GUI to try to delete a file, and verify that you get a notification that you don't have permission to do that
4. Use the submit data page to upload the same data folder twice, and add comments so that you can identify these uploads in the metadata browser
5. Go to the metadata browser, find your two uploads from the previous step, and verify their data folders use different prefixes and each contain all of the expected data
6. Deploy your instance with a new user password using  `MINIMO_USER_PASSWORD=yourpassword docker-compose up -d --build --force-recreate`
7. Hit the data browser directly at `https://minimo.localhost/minio/data` and verify that you can log in with credentials `minimouser` and `yourpassword`
